File: LeetCode/112WordLadder/WordLadder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        // BFS -> Time:O(n^2), Space:O(n^2)
        set<string> wordSet(wordList.begin(),wordList.end());
        if (wordSet.find(endWord) == wordSet.end()) {
            return 0;
        }
        queue<map<string,int>> q;
        int level = 1;
        map<string,int> m;
        m[beginWord] = level;
        q.push(m);
        set<string> visited;
        visited.insert(beginWord);
        while (!q.empty()) {
            map<string,int> node = q.front();
            string word = node.begin()->first;
            level = node.begin()->second;
            q.pop();
            if (word == endWord) {
                return level;
            }
            for (int i = 0; i < word.size(); i++) {
                for (int j = 97; j < 123; j++) {
                    string modWord = word.substr(0, i) + char(j) + word.substr(i+1, word.size()-i);
                    if (wordSet.find(modWord) != wordSet.end() && 
                    visited.find(modWord) == visited.end()) {
                        visited.insert(modWord);
                        node.clear();
                        node[modWord] = level + 1;
                        q.push(node);
                    }
                }
            }
        }
        return 0;
    }
};

File: LeetCode/112WordLadder/WordLadder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        # BFS -> Time:O(n^2), Space:O(n^2)
        word_set = set(wordList)
        if endWord not in word_set:
            return 0
        q = []
        visited = set()
        q.append([beginWord, 1])
        visited.add(beginWord)
        while q:
            word, level = q.pop(0)
            if word == endWord:
                return level
            for i in range(len(word)):
                for j in range(ord("a"), ord("z")+1):
                    mod_word = word[:i] + chr(j) + word[i+1:]
                    if mod_word in word_set and mod_word not in visited:
                        q.append([mod_word, level + 1])
                        visited.add(mod_word)
        return 0
